include/hybridmap/AodScheduler.hpp

```diff
@@ -0,0 +1,302 @@
+//
+
+#pragma once
+
+#include "Definitions.hpp"
+#include "QuantumComputation.hpp"
+#include "cstdint"
+#include "hybridmap/NeutralAtomArchitecture.hpp"
+#include "hybridmap/NeutralAtomDefinitions.hpp"
+#include "hybridmap/NeutralAtomUtils.hpp"
+#include "operations/AodOperation.hpp"
+#include "operations/OpType.hpp"
+#include "utility"
+#include "vector"
+
+namespace qc {
+// Possible types two Move combination can be combined to
+enum class ActivationMergeType : uint8_t { Impossible, Trivial, Merge, Append };
+
+/**
+ * @brief Class to schedule AOD movements on a neutral atom architecture
+ * @details The scheduler takes a quantum circuit containing abstract move
+ * operations and tries to merge them into parallel AO movements. It also
+ * manages the small offset movements required while loading or unloading of
+ * AODs.
+ */
+class AodScheduler {
+protected:
+  /**
+   * @brief Struct to store information about specific AOD activations.
+   * @details It contains:
+   * - the offset moves in x and y direction
+   * - the actual moves
+   */
+  struct AodActivationHelper {
+    /**
+     * @brief Describes a single AOD movement in either x or y direction
+     * @details It contains:
+     * - the initial position of the AOD
+     * - the delta of the AOD movement
+     * - the size of the offset of the AOD movement
+     */
+    struct AodMove {
+      // start of the move
+      uint32_t init;
+      // need offset move to avoid crossing
+      int32_t offset;
+      // delta of the actual move
+      fp delta;
+
+      AodMove(uint32_t init, fp delta, int32_t offset)
+          : init(init), offset(offset), delta(delta) {}
+    };
+    /**
+     * @brief Manages the activation of an atom using an AOD.
+     * @details The same struct is used also to deactivate the AOD, just
+     * reversed.
+     */
+    struct AodActivation {
+      // first: x, second: delta x, third: offset x
+      std::vector<AodMove*> activateXs;
+      std::vector<AodMove*> activateYs;
+      std::vector<AtomMove> moves;
+
+      AodActivation(const AodMove& activateXs, const AodMove& activateYs,
+                    const AtomMove& move)
+          : activateXs({new AodMove(activateXs)}),
+            activateYs({new AodMove(activateYs)}), moves({move}) {}
+      AodActivation(const Dimension dim, const AodMove& activate,
+                    const AtomMove& move)
+          : moves({move}) {
+        if (dim == Dimension::X) {
+          activateXs = {new AodMove(activate)};
+        } else {
+          activateYs = {new AodMove(activate)};
+        }
+      }
+
+      [[nodiscard]] std::vector<AodMove*> getActivates(Dimension dim) const {
+        if (dim == Dimension::X) {
+          return activateXs;
+        }
+        return activateYs;
+      }
+    };
+
+    // AODScheduler
+    // NeutralAtomArchitecture to call necessary hardware information
+    NeutralAtomArchitecture    arch;
```

Comment:
Another copy of the architecture? Might be an architectural decision to rethink. Probably the easiest fix is to replace this with a `NeutralAtomArchitecture*` and use `&arch` in the constructor.

Reply:
This is indeed not ideal, but the AodScheduler needs some information on the atom spacing which is provided by the Architecture class.
Regarding the pointer, I thought that the fact that I use std::move(arch) should avoid an additional copy.

Reply:
Unfortunately, the initialization of the member will still create a copy. In case of passing by reference, the copy happens in the constructor itself. In case of the move version, the architecture itself is passed by value, hence, creating a copy.

But I believe the solution here is rather straightforward by simply replacing the Architecture member with a pointer to an architecture. Happy to try that out tomorrow, if you don't feel comfortable with it.

Reply:
Ah, then I had a wrong understanding of the passing by reference. Can then also change to the pointer version, no problem

Reply:
Fixed it with const ref members. Is this how one should do this in general, or should I just always pass the architecture by argument?

Reply:
I'd have to see in the actual code to check (just in my phone right now), but in general one should avoid reference members in a class (i.e., `NeutralAtomArchitecture& arch`).
The main reasoning for that is written down here: https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines/avoid-const-or-ref-data-members.html
What you would want to do is to make the member a pointer to an architecture (i.e., `NeutralAtomArchitecture* arch`), have the constructor take a reference `NeutralAtomArchitecture& architecture`, and use `arch(&architecture)` to set the pointer member in the constructor based on the address (`&`) of the passed architecture.

Reply:
For cases where I don't change the arch and I don't want to copy the object I kept the ref members, in the other cases I used a pointer.